### PR TITLE
SortingAnalyzer: `get_extension_default_params`

### PR DESCRIPTION
**Pull request by alejoe91**
and `get_computable_extensions()`

@samuelgarcia @h-mayorquin @zm711 couldn't come up with a better name for the latter. Any suggestion?

At one point, we will make the switch and define proper pydantic models (at least for extension, sorters, quality metrics, template metrics params)

**NOTE:** I changed the `unit_locations` `**method_kwargs` to a `method_kwargs` dict, which is better for inspection

### CODE REVIEW

**Comment by zm711** [comment]
> and get_computable_extensions()

I'm not opposed to `get_computable_extensions`. Since the main syntax of the analyzer is `analyzer.compute(...)` having a get what is computable makes sense. `computable` is not the most discoverable of words, though so I can understand not liking that. I think the more common wording would be `get_possible_extensions` in that because they can be computed they are possible. But we already use `possible_extensions` elsewhere so I don't know if we want to overload.

So for me it comes back to choose an easy intuitive word for beginning users (possible) or choose a more precise accurate word for advanced/dev users (computable). Interested in what the others think!

**Review by h-mayorquin — APPROVED** [review]
This looks OK to me.
It is a verb, it is specific and it does what is says. But for some suggestions:

"get _possible_extenssions" works as well the same as  "get_available_extensions". It could also be an attribute instead of a getter. 

Random suggestions that could be useful just because I thought about them:
* What about an attribute?
* Should it be a static method so the user can query which extensions are available without need to instantiate the class? (mabye there is another way to do this that I just don't know about?)
* In the same vein as the last, should there be an attribute that can be imported from top level called "available_sorting_analyzer_extensions`?

Not very well though off, just throwing some random suggetions because sometimes is useful.

**Comment by alejoe91** [comment]
I like the idea of moving the two functions at the module level, as one might want to explore extensions and params before instantiating the analyzer.

What do you think abuot `si.get_available_analyzer_extensions()` and `si.get_analyzer_extension_params()` (I would drop the default here)?

**Comment by zm711** [comment]
> and si.get_analyzer_extension_params() (I would drop the default here)?

For this one you would take in the `extension_name` and return params for just one? I think returning all default params would be confusing, but I like the idea of allowing someone to easily grab the defaults for an extension at a time.

**Comment by alejoe91** [comment]
> > and si.get_analyzer_extension_params() (I would drop the default here)?
> 
> For this one you would take in the `extension_name` and return params for just one? I think returning all default params would be confusing, but I like the idea of allowing someone to easily grab the defaults for an extension at a time.

Yep, same as It Is now

**Comment by zm711** [comment]
> What do you think abuot si.get_available_analyzer_extensions() and si.get_analyzer_extension_params() (I would drop the default here)?

Then this sounds good to me :)

**Comment by alejoe91** [comment]
Done!

File: src/spikeinterface/core/sortinganalyzer.py [hunk 1]
```diff
@@ -1091,6 +1113,39 @@ def get_extension_class(extension_name: str, auto_import=True):
...
 
 
+def get_available_analyzer_extensions():
+    """
+    Get all extensions that can be computed by the analyzer.
+    """
+    return list(_builtin_extensions.keys())
+
+
+def get_default_anlyzer_extension_params(extension_name: str):
```
Comment (zm711) [inline src/spikeinterface/core/sortinganalyzer.py:1123]:
```suggestion
def get_default_analyzer_extension_params(extension_name: str):
```

File: src/spikeinterface/core/__init__.py [hunk 2]
```diff
@@ -144,7 +144,15 @@
...
 
 # SortingAnalyzer and AnalyzerExtension
-from .sortinganalyzer import SortingAnalyzer, AnalyzerExtension, create_sorting_analyzer, load_sorting_analyzer
+from .sortinganalyzer import (
+    SortingAnalyzer,
+    AnalyzerExtension,
+    create_sorting_analyzer,
+    load_sorting_analyzer,
+    get_available_analyzer_extensions,
+    get_default_anlyzer_extension_params,
```
Comment (zm711) [inline src/spikeinterface/core/__init__.py:153]:
```suggestion
    get_default_analyzer_extension_params,
```

File: src/spikeinterface/core/sortinganalyzer.py [hunk 3]
```diff
@@ -1025,6 +1025,28 @@ def has_extension(self, extension_name: str) -> bool:
...
+        ----------
+        extension_name: str
+            The extension name
+
+        Returns
+        -------
+        default_params: dict
+            The default parameters for the extension
+        """
+        return get_default_anlyzer_extension_params(extension_name)
```
Comment (zm711) [inline src/spikeinterface/core/sortinganalyzer.py:1048]:
```suggestion
        return get_default_analyzer_extension_params(extension_name)
```

File: src/spikeinterface/core/tests/test_sortinganalyzer.py [hunk 4]
```diff
@@ -143,6 +148,24 @@ def _check_sorting_analyzers(sorting_analyzer, original_sorting):
...
 
+def test_extension_params():
+    from spikeinterface.core.sortinganalyzer import _builtin_extensions
+
+    computable_extension = get_available_analyzer_extensions()
+
+    for ext, mod in _builtin_extensions.items():
+        assert ext in computable_extension
+        if mod == "spikeinterface.core":
+            default_params = get_default_anlyzer_extension_params(ext)
```
Comment (zm711) [inline src/spikeinterface/core/tests/test_sortinganalyzer.py:159]:
```suggestion
            default_params = get_default_analyzer_extension_params(ext)
```

File: src/spikeinterface/core/tests/test_sortinganalyzer.py [hunk 5]
```diff
@@ -143,6 +148,24 @@ def _check_sorting_analyzers(sorting_analyzer, original_sorting):
...
+            print(ext, default_params)
+        else:
+            try:
+                default_params = get_default_anlyzer_extension_params(ext)
```
Comment (zm711) [inline src/spikeinterface/core/tests/test_sortinganalyzer.py:163]:
```suggestion
                default_params = get_default_analyzer_extension_params(ext)
```

**Review by zm711 — COMMENTED** [review]
I think I got them all :) but anlyzer -> analyzer.

**Comment by zm711** [comment]
Based on failing tests I must've missed one sorry!

File: src/spikeinterface/core/tests/test_sortinganalyzer.py [hunk 6]
```diff
@@ -4,7 +4,12 @@
 import shutil
 
 from spikeinterface.core import generate_ground_truth_recording
-from spikeinterface.core import SortingAnalyzer, create_sorting_analyzer, load_sorting_analyzer
+from spikeinterface.core import (
+    create_sorting_analyzer,
+    load_sorting_analyzer,
+    get_available_analyzer_extensions,
+    get_default_anlyzer_extension_params,
```
Comment (zm711) [inline src/spikeinterface/core/tests/test_sortinganalyzer.py:11]:
```suggestion
    get_default_analyzer_extension_params,
```

File: src/spikeinterface/postprocessing/unit_localization.py [hunk 7]
```diff
@@ -59,7 +57,7 @@ class ComputeUnitLocations(AnalyzerExtension):
     def __init__(self, sorting_analyzer):
         AnalyzerExtension.__init__(self, sorting_analyzer)
 
-    def _set_params(self, method="monopolar_triangulation", **method_kwargs):
+    def _set_params(self, method="monopolar_triangulation", method_kwargs={}):
```
Comment (samuelgarcia) [inline src/spikeinterface/postprocessing/unit_localization.py:60]: I am not agree with this.

Comment (samuelgarcia) [reply src/spikeinterface/postprocessing/unit_localization.py:60]: Can we revert this ?

File: src/spikeinterface/core/sortinganalyzer.py [hunk 8]
```diff
@@ -1091,6 +1113,39 @@ def get_extension_class(extension_name: str, auto_import=True):
     return ext_class
 
 
+def get_available_analyzer_extensions():
+    """
+    Get all extensions that can be computed by the analyzer.
+    """
+    return list(_builtin_extensions.keys())
```
Comment (samuelgarcia) [inline src/spikeinterface/core/sortinganalyzer.py:1120]: This should this + the one that are register. We can have extension outside!

**Comment by alejoe91** [comment]
> I am OK with the new method but I am not agree with the signature change inunit_locaization. This make non backward compatible witg previous versions. We should have discuss this before this super fast merge, I think. No ?

That unifies the params with the spike locations. In general, having **method_kwargs in the set params is bad IMO.

Not that it doesn't really break back-compatibility. For results that are already saved as a `WaveformExtractor`, the params don't do anything since we can't recompute stuff with it.

**Pull request merged**